samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java

```diff
@@ -144,122 +58,108 @@ public StreamGraphImpl(ApplicationRunner runner, Config config) {
   }
 
   @Override
-  public <K, V, M extends MessageEnvelope<K, V>> MessageStream<M> createInStream(StreamSpec streamSpec, Serde<K> keySerde, Serde<V> msgSerde) {
-    if (!this.inStreams.containsKey(streamSpec.getId())) {
-      this.inStreams.putIfAbsent(streamSpec.getId(), new InputStreamImpl<K, V, M>(this, streamSpec, keySerde, msgSerde));
-    }
-    return this.inStreams.get(streamSpec.getId());
+  public <K, V, M> MessageStream<M> createInStream(StreamSpec streamSpec,
+      BiFunction<K, V, M> msgBuilder, Serde<K> keySerde, Serde<V> msgSerde) {
+    return inStreams.computeIfAbsent(streamSpec.getId(),
+        streamId -> new InputStreamImpl<>(this, streamSpec, msgBuilder, keySerde, msgSerde));
+  }
+
+  @Override
+  public <K, V, M> MessageStream<M> createOutStream(StreamSpec streamSpec,
+      Function<M, K> keyExtractor, Function<M, V> msgExtractor,
+      Serde<K> keySerde, Serde<V> msgSerde) {
+    return outStreams.computeIfAbsent(streamSpec.getId(),
+        streamId -> new OutputStreamImpl<>(this, streamSpec, keyExtractor, msgExtractor, keySerde, msgSerde));
+  }
+
+  @Override
+  public StreamGraph withContextManager(ContextManager manager) {
+    this.contextManager = manager;
+    return this;
   }
 
   /**
-   * Helper method to be used by {@link MessageStreamImpl} class
+   * Internal helper method to create an intermediate stream for an operator.
    *
-   * @param streamSpec  the {@link StreamSpec} object defining the {@link SystemStream} as the output
-   * @param <M>  the type of {@link MessageEnvelope}s in the output {@link SystemStream}
-   * @return  the {@link MessageStreamImpl} object
+   * @param streamName the name of the stream to be created (will be prefixed with job name and id)
+   * @param keyExtractor the function to extract the output key from the input message
+   * @param msgExtractor the function to extract the output message from the input message
+   * @param msgBuilder the function to convert the incoming key and message to the output message
+   * @param <K> the type of key in the incoming message
+   * @param <V> the type of message in the incoming message
+   * @param <M> the type of messages in the intermediate {@link MessageStream}
+   * @return  the output {@link MessageStreamImpl} for the re-partitioned stream
    */
-  @Override
-  public <K, V, M extends MessageEnvelope<K, V>> OutputStream<M> createOutStream(StreamSpec streamSpec, Serde<K> keySerde, Serde<V> msgSerde) {
-    if (!this.outStreams.containsKey(streamSpec.getId())) {
-      this.outStreams.putIfAbsent(streamSpec.getId(), new OutputStreamImpl<K, V, M>(this, streamSpec, keySerde, msgSerde));
-    }
-    return this.outStreams.get(streamSpec.getId());
+  <K, V, M> MessageStreamImpl<M> createIntStream(String streamName, Function<M, K> keyExtractor, Function<M, V> msgExtractor,
+      BiFunction<K, V, M> msgBuilder) {
+    String streamId = String.format("%s-%s-%s",
```

Comment:
As an aside, when we move application runner out of SamzaContainer (SAMZA-1159 phase 2, depends on SAMZA-1118), ExecutionPlanner should be the one adding the job/app name and id to the operator name instead of MessageStreamImpl.